File: src/disassembler.cpp
#include <Zycore/Status.h>
#include <Zydis/Disassembler.h>
#include <Zydis/Zydis.h>

#include <cstddef>
#include <libxdb/disassembler.hpp>

namespace xdb {

std::vector<disassembler::instruction> disassembler::disassemble(
    std::size_t n_instructions, std::optional<virt_addr> address) {
    std::vector<instruction> res;
    res.reserve(n_instructions);

    // Default address is the current program counter
    if (!address) {
        address = process_->get_pc();
    }

    const std::size_t MAX_LENGTH_OF_X86_64_INSTRUCTION = 15;
    auto code = process_->read_memory_without_traps(
        address.value(), n_instructions * MAX_LENGTH_OF_X86_64_INSTRUCTION);
    std::span<const std::byte> code_span(code.data(), code.size());

    ZydisDisassembledInstruction instr;
    while (n_instructions > 0 && !code_span.empty() &&
           ZYAN_SUCCESS(ZydisDisassembleIntel(ZYDIS_MACHINE_MODE_LONG_64,
                                              address->addr(), code_span.data(),
                                              code_span.size(), &instr))) {
        res.emplace_back(*address, instr.text);
        --n_instructions;
        *address += instr.info.length;
        code_span = code_span.subspan(instr.info.length);
    }

    return res;
}

}  // namespace xdb

File: src/disassembler.rs
use iced_x86::{Decoder, DecoderOptions, FlowControl, Formatter, IntelFormatter};

use crate::error::Result;
use crate::process::Process;
use crate::types::VirtAddr;

const MAX_X86_64_INSTRUCTION_LEN: usize = 15;

#[derive(Clone, Debug)]
pub struct Instruction {
    pub address: VirtAddr,
    pub text: String,
    pub is_call: bool,
}

pub struct Disassembler<'a> {
    process: &'a Process,
}

impl<'a> Disassembler<'a> {
    pub fn new(process: &'a Process) -> Self {
        Self { process }
    }

    pub fn disassemble(&self, count: usize, address: Option<VirtAddr>) -> Result<Vec<Instruction>> {
        // Default address is the current program counter
        let start = match address {
            Some(addr) => addr,
            None => self.process.get_pc()?,
        };

        let code = self
            .process
            .read_memory_without_traps(start, count * MAX_X86_64_INSTRUCTION_LEN)?;
        let mut decoder = Decoder::with_ip(64, &code, start.addr(), DecoderOptions::NONE);
        let mut formatter = IntelFormatter::new();

        let mut result = Vec::with_capacity(count);
        let mut text = String::new();
        while result.len() < count && decoder.can_decode() {
            let instruction = decoder.decode();
            text.clear();
            formatter.format(&instruction, &mut text);
            let is_call = matches!(
                instruction.flow_control(),
                FlowControl::Call | FlowControl::IndirectCall
            );
            result.push(Instruction {
                address: VirtAddr(instruction.ip()),
                text: text.clone(),
                is_call,
            });
        }
        Ok(result)
    }
}
